crates/test/mock-consensus/src/builder/init_chain.rs

```diff
@@ -0,0 +1,94 @@
+use {
+    super::*,
+    anyhow::{anyhow, bail},
+    std::time,
+    tap::TapFallible,
+    tendermint::{
+        block,
+        consensus::{
+            self,
+            params::{AbciParams, ValidatorParams, VersionParams},
+        },
+        evidence,
+        v0_37::abci::{ConsensusRequest, ConsensusResponse},
+    },
+    tower::{BoxError, Service, ServiceExt},
+    tracing::{debug, error},
+};
+
+impl Builder {
+    /// Consumes this builder, using the provided consensus service.
+    pub async fn init_chain<C>(self, mut consensus: C) -> Result<TestNode<C>, anyhow::Error>
+    where
+        C: Service<ConsensusRequest, Response = ConsensusResponse, Error = BoxError>
+            + Send
+            + Clone
+            + 'static,
+        C::Future: Send + 'static,
+        C::Error: Sized,
+    {
+        use tendermint::v0_37::abci::response;
+
+        let request = Self::init_chain_request();
+        let service = consensus
+            .ready()
+            .await
+            .tap_err(|error| error!(?error, "failed waiting for consensus service"))
+            .map_err(|_| anyhow!("failed waiting for consensus service"))?;
+
+        let response::InitChain { app_hash, .. } = match service
+            .call(request)
+            .await
+            .tap_ok(|resp| debug!(?resp, "received response from consensus service"))
+            .tap_err(|error| error!(?error, "consensus service returned error"))
+            .map_err(|_| anyhow!("consensus service returned error"))?
+        {
+            ConsensusResponse::InitChain(resp) => resp,
+            response => {
+                error!(?response, "unexpected InitChain response");
+                bail!("unexpected InitChain response");
+            }
+        };
```

Comment:
tioli: i think this could be slightly cleaner if it were a `let ... else { bail!() }` structure

Reply:
i opted for this form so we have the chance to log what we _did_ get in the unexpected case, though i'm a fan of `let .. else` is 99% of cases :slightly_smiling_face: